plugins/base/src/main/java/edu/wpi/first/shuffleboard/plugin/base/widget/GraphWidget.java

```diff
@@ -282,111 +247,110 @@ private void updateFromArraySource(DataSource<double[]> source) {
-        newSeries.setName(source.getName() + "[" + i + "]"); // eg "array[0]", "array[1]", etc
+        DoubleDataSet newSeries = new DoubleDataSet(source.getName() + "[" + i + "]");
         series.add(newSeries);
-        realData.put(newSeries, new SimpleData());
         visibleSeries.computeIfAbsent(newSeries, createVisibleProperty);
       }
     }
     return series;
   }
 
-  private void updateBounds(long elapsed) {
-    xAxis.setUpperBound(elapsed);
-    removeInvisibleData();
-  }
-
   private void update() {
     FxUtils.runOnFxThread(() -> {
-      if (chart.getData().isEmpty()) {
-        return;
+      // Data is only pushed to the graph via listeners, so this prevents the graph
+      // from staying still during a period of no updates.
+      for (var source : numberSeriesMap.keySet()) {
+        numberChangeLister.changed(source.dataProperty(), null, source.getData());
       }
-      synchronized (queueLock) {
-        queuedData.forEach((series, queuedData) -> series.getData().addAll(queuedData));
-        queuedData.forEach((series, queuedData) -> queuedData.clear());
-        OptionalLong maxX = chart.getData().stream()
-            .map(Series::getData)
-            .filter(d -> !d.isEmpty())
-            .map(d -> d.get(d.size() - 1))
-            .map(Data::getXValue)
-            .mapToLong(Number::longValue)
-            .max();
-        if (maxX.isPresent()) {
-          updateBounds(maxX.getAsLong());
+
+      for (var source : arraySeriesMap.keySet()) {
+        numberArrayChangeListener.changed(source.dataProperty(), null, source.getData());
+      }
+
+      // This code actually rerenders the graph.
+
+      OptionalDouble globalMax = OptionalDouble.empty();
+      for (DataSet s : chart.getDatasets()) {
+        var doubleDataSet = (DoubleDataSet) s;
+
+        OptionalDouble dataSetMax = doubleDataSet.lock().readLockGuard(() -> {
+
+          if (doubleDataSet.getDataCount(DataSet.DIM_X) == 0) {
+            return OptionalDouble.empty();
+          }
+
+          double[] xValues = doubleDataSet.getValues(DataSet.DIM_X);
+          return OptionalDouble.of(xValues[doubleDataSet.getDataCount(DataSet.DIM_X) - 1]);
+        });
+
+        if (dataSetMax.isPresent()) {
+          if (globalMax.isPresent() && dataSetMax.getAsDouble() > globalMax.getAsDouble()) {
+            globalMax = dataSetMax;
+          } else if (globalMax.isEmpty()) {
+            globalMax = dataSetMax;
+          }
+        }
+
+        if (dataSetMax.isPresent() && autoScrollToggle.isSelected()) {
+          xAxis.maxProperty().set(dataSetMax.getAsDouble());
+          xAxis.minProperty().bind(xAxis.maxProperty().subtract(visibleTime.multiply(1e3)));
+          doubleDataSet.fireInvalidated(null);
+        } else {
+          xAxis.maxProperty().unbind();
+          xAxis.minProperty().unbind();
         }
       }
+
+      if (autoScrollToggle.isSelected() && globalMax.isPresent()) {
```

Comment:
Consider putting `autoScrollToggle.isSelected()` into a helper method:

```java
private boolean isAutoScroll() {
  return autoScrollToggle.isSelected();
}
```

Alternatively, if autoscroll is made into a context menu `Action`, you'd need this method anyway to read from the boolean property of the setting.